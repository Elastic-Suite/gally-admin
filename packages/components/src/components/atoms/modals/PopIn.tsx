import React, { PropsWithChildren, ReactNode, useState } from 'react'
import { Box, DialogActions, DialogContent, DialogTitle } from '@mui/material'
import {
  CustomClose,
  CustomDialog,
  CustomTitle,
  IPropsCustomDialog,
} from './PopIn.styled'
import IonIcon from '../IonIcon/IonIcon'
import Button from '../buttons/Button'
import { useLog } from '../../../hooks'

interface IProps extends Omit<IPropsCustomDialog, 'open'> {
  confirmationPopIn?: boolean
  triggerElement: ReactNode
  actions?: ReactNode
  titlePopIn?: string
  onConfirm?: () => void | Promise<void>
  cancelName?: string
  confirmName?: string
  loading?: boolean
}

function PopIn({
  confirmationPopIn,
  triggerElement,
  titlePopIn,
  actions,
  children,
  position,
  styles,
  onConfirm,
  cancelName,
  confirmName,
  loading,
}: PropsWithChildren<IProps>): JSX.Element {
  const [open, setOpen] = useState(false)
  const log = useLog()

  function handleClickOpen(): void {
    setOpen(true)
  }

  function handleClose(): void {
    setOpen(false)
  }

  async function handleConfirm(): Promise<void> {
    if (typeof onConfirm === 'function') {
      try {
        const result = onConfirm()
        if (result instanceof Promise) {
          return await result
        }
      } catch (error) {
        return log(error)
      }
    }
    setOpen(false)
  }

  return (
    <div style={{ position: 'relative' }}>
      <Box onClick={handleClickOpen}>{triggerElement}</Box>

      <CustomDialog
        open={open}
        onClose={handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
        position={position}
        styles={
          confirmationPopIn
            ? {
                actions: {
                  padding: '0 32px 32px 32px',
                  justifyContent: 'center',
                  gap: 1,
                },
              }
            : styles
        }
      >
        <CustomClose onClick={handleClose}>
          <IonIcon name="close" style={{ fontSize: '17.85px' }} />
        </CustomClose>
        {!confirmationPopIn ? (
          <>
            {titlePopIn ? <DialogTitle> {titlePopIn} </DialogTitle> : null}
            <DialogContent>{children}</DialogContent>
            <DialogActions> {actions} </DialogActions>
          </>
        ) : (
          <>
            <CustomTitle> {titlePopIn} </CustomTitle>
            <DialogActions>
              <Box>
                <Button onClick={handleClose} display="tertiary" size="large">
                  {cancelName}
                </Button>
              </Box>
              <Box>
                <Button onClick={handleConfirm} loading={loading} size="large">
                  {confirmName}
                </Button>
              </Box>
            </DialogActions>
          </>
        )}
      </CustomDialog>
    </div>
  )
}

export default PopIn
