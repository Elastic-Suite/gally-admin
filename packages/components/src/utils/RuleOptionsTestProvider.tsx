import React, { ReactNode } from 'react'
import {
  IRuleEngineOperators,
  RuleAttributeType,
  getIri,
} from '@elastic-suite/gally-admin-shared'

import ruleEngineOperator from '../../public/mocks/rule_engine_operators.json'
import sourceFields from '../../public/mocks/source_fields.json'
import sourceFieldLabels from '../../public/mocks/source_field_labels.json'

import RuleOptionsProvider from '../components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(getIri('source_fields', field.id))
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator as IRuleEngineOperators}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
