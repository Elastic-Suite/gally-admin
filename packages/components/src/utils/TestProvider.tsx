import React, { ReactNode } from 'react'

import catalogs from '../../public/mocks/catalog.json'
import metadata from '../../public/mocks/metadata.json'

import { catalogContext, metadataContext } from '../contexts'
import { useCatalogs } from '../hooks'

import OptionsProvider from '../components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { children } = props
  const contextValueCatalog = useCatalogs(catalogs['hydra:member'])
  const contextValueMetadata = metadata['hydra:member']

  return (
    <OptionsProvider>
      <catalogContext.Provider value={contextValueCatalog}>
        <metadataContext.Provider value={contextValueMetadata}>
          {children}
        </metadataContext.Provider>
      </catalogContext.Provider>
    </OptionsProvider>
  )
}

export default TestProvider
