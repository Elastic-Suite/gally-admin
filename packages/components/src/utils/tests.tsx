import React, { ReactNode } from 'react'
import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { Bundle, api } from '@elastic-suite/gally-admin-shared'

import { AppStore, RootState, setupStore } from '../store'

import AppProvider from '../components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: {
      'base_url/media': 'https://localhost/media/catalog/product/',
    },
    metadata: [
      {
        '@id': '/metadata/3',
        '@type': 'Metadata',
        id: 3,
        entity: 'product',
        sourceFields: ['/source_fields/217', '/source_fields/253'],
      },
      {
        '@id': '/metadata/4',
        '@type': 'Metadata',
        id: 4,
        entity: 'category',
        sourceFields: ['/source_fields/254', '/source_fields/286'],
      },
    ],
  },
  user: {
    requestedPath: '',
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
