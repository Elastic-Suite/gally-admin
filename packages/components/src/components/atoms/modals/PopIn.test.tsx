import React from 'react'

import { renderWithProviders } from '../../../utils/tests'

import Button from '../buttons/Button'

import PopIn from './PopIn'

describe('PopIn match snapshot', () => {
  it('PopIn simple', () => {
    const { container } = renderWithProviders(
      <PopIn
        position="center"
        confirmationPopIn
        triggerElement={<Button size="large">Click on me !</Button>}
        cancelName="Cancel"
        confirmName="Confirm"
        titlePopIn="Hello World"
      />
    )
    expect(container).toMatchSnapshot()
  })
})
