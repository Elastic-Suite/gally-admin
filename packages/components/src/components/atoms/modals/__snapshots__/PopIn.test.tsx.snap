// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopIn match snapshot PopIn simple 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-zwkw2d-MuiButtonBase-root-MuiButton-root"
      data-testid="button"
      tabindex="0"
      type="button"
    >
      Click on me !
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
